fdb/fdb.go

```diff
@@ -0,0 +1,262 @@
+package fdb
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/apple/foundationdb/bindings/go/src/fdb"
+	"github.com/apple/foundationdb/bindings/go/src/fdb/directory"
+	"github.com/apple/foundationdb/bindings/go/src/fdb/tuple"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipni/dhstore"
+	"github.com/multiformats/go-multicodec"
+	"github.com/multiformats/go-multihash"
+	"lukechampine.com/blake3"
+)
+
+var (
+	_ dhstore.DHStore = (*FDBDHStore)(nil)
+
+	logger                   = logging.Logger("store/fdb")
+	fdbHasherPool            sync.Pool
+	errMultihashDigestLength = errors.New("multihash digest must be exactly 32 bytes long")
+	errMetadataKeyTooLong    = errors.New("key must be at most 32 bytes long")
+)
+
+const (
+	fdbMaxValueBytes      = 100_000 // 100 KB
+	fdbBlake3HashLength   = 32
+	fdbMultihashDirectory = "mh"
+	fdbMetadataDirectory  = "md"
+	// fdbMaxKeyPrefixLen is the threshold at which key prefixes are hashed if they are larger. Otherwise,
+	// the prefix is used as is. When used in the context of metadata keys, it represents the max accepted
+	// length for metadata key.
+	fdbMaxKeyPrefixLen = 32
+)
+
+type FDBDHStore struct {
+	db fdb.Database
+}
+
+func init() {
+	fdbHasherPool.New = func() any {
+		return blake3.New(fdbBlake3HashLength, nil)
+	}
+}
+
+func NewFDBDHStore(o ...Option) (*FDBDHStore, error) {
+	opts, err := newOptions(o...)
+	if err != nil {
+		return nil, err
+	}
+	if err := fdb.APIVersion(opts.apiVersion); err != nil {
+		return nil, err
+	}
+	db, err := fdb.OpenDatabase(opts.clusterFile)
+	if err != nil {
+		panic(err)
+	}
+	return &FDBDHStore{
+		db: db,
+	}, nil
+}
+
+func (f *FDBDHStore) MergeIndex(mh multihash.Multihash, vk dhstore.EncryptedValueKey) error {
+	dmh, err := multihash.Decode(mh)
+	if err != nil {
+		return dhstore.ErrMultihashDecode{Err: err, Mh: mh}
+	}
+	if multicodec.Code(dmh.Code) != multicodec.DblSha2_256 {
+		return dhstore.ErrUnsupportedMulticodecCode{Code: multicodec.Code(dmh.Code)}
+	}
+	if dmh.Length != 32 {
+		return dhstore.ErrMultihashDecode{Err: errMultihashDigestLength, Mh: mh}
+	}
+	if len(vk) > fdbMaxValueBytes {
+		return fmt.Errorf("value key cannot be larger than 100 KB, got: %d", len(vk))
+	}
+	_, err = f.db.Transact(func(transaction fdb.Transaction) (any, error) {
```

Comment:
This is doing a separate transaction for each individual multihash. Seems like the transaction should be wrapped around a batch of multihashes. Seems like separate transactions would cause separate db connections, and defeat what we want to use FDB for in the first place.

Reply:
I read on this a bit and have some ideas.
 I'd like to get some basic numbers out before spending more time. 
 
 TL;DR 
* the recommendation is to use concurrent goroutines each open a transaction.
* In our case repeated multihashes are relatively rare so writing the entire batch in one transaction seems reasonable.
* this needs interface changes across dhstore which i'd like to hold off since it causes conflicts with other open PRs.

Reply:
Also see:  [FDB Read/Write Path](https://apple.github.io/foundationdb/read-write-path.html?highlight=batch)

> Client. It is a library, an FDB application uses, to access the database. It exposes the transaction concept to applications. Client in FDB is a fat client that does multiple complex operations: (1) It calculates read and write conflict ranges for transactions; (2) it batches a transaction’s operations and send them all together at commit for better throughput; (3) it automatically retries failed transactions.

Reply:
OK, proceed.